File: frontend/mobile/lib/models/responses/sector_response.dart
import '../api_enum.dart';
import '../ticketing_mode.dart';
import 'ticket_type_response.dart';

class SectorResponse {
  final String id;
  final String productId;
  final String name;
  final int capacity;
  final double price;
  final String status;
  final TicketingMode ticketingMode;
  final int? periodYear;
  final int? periodMonth;
  final DateTime createdAt;
  final List<TicketTypeResponse> ticketTypes;

  const SectorResponse({
    required this.id,
    required this.productId,
    required this.name,
    required this.capacity,
    required this.price,
    required this.status,
    required this.ticketingMode,
    this.periodYear,
    this.periodMonth,
    required this.createdAt,
    required this.ticketTypes,
  });

  factory SectorResponse.fromJson(Map<String, dynamic> json) {
    if (json['id'] == null) throw const FormatException('Missing id in payload');
    if (json['productId'] == null) throw const FormatException('Missing productId in payload');
    if (json['name'] == null) throw const FormatException('Missing name in payload');
    if (json['price'] == null) throw const FormatException('Missing price in payload');

    return SectorResponse(
      id: json['id'] as String,
      productId: json['productId'] as String,
      name: json['name'] as String,
      capacity: json['capacity'] as int? ?? 0,
      price: (json['price'] as num).toDouble(),
      status: publishStatusFromJson(json['status']),
      ticketingMode: ticketingModeFromJson(json['ticketingMode']),
      periodYear: json['periodYear'] as int?,
      periodMonth: json['periodMonth'] as int?,
      createdAt: DateTime.tryParse(json['createdAt'] as String? ?? '') ?? DateTime.now(),
      ticketTypes:
          (json['ticketTypes'] as List? ?? []).map((e) => TicketTypeResponse.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }
}

File: backend/services/ticketing/eTicketing.Ticketing.Business/Sectors/Dtos/SectorResponse.cs
using eTicketing.Contracts.Pagination;
using eTicketing.Contracts.Persistence;

namespace eTicketing.Ticketing.Business.Sectors;

public record SectorResponse(
    Guid Id,
    Guid ProductId,
    string Name,
    int Capacity,
    decimal Price,
    PublishStatus Status,
    TicketingMode TicketingMode,
    int? PeriodYear,
    int? PeriodMonth,
    DateTime CreatedAt,
    IReadOnlyList<TicketTypeResponse> TicketTypes,
    /// <summary>How many admissions are still on sale, read off the live Redis counter — what the
    /// buying clients need to grey a sold-out option out instead of letting someone pick it and
    /// collect a 409 at hold time. Populated on the public buyer path only (GET /sectors); the
    /// organizer/staff lists leave it null rather than pay a Redis round-trip per row for a number
    /// their screens don't show. Also null for a DailyEntry sector queried without a date, because
    /// that mode counts capacity per (Sector, date) and a single number would be a lie.</summary>
    int? RemainingCapacity = null)
{
    /// <summary>True only when availability is actually known and exhausted. An unknown
    /// (null) remaining must never read as sold out — that would hide sectors that are on sale.
    /// </summary>
    public bool IsSoldOut => RemainingCapacity == 0;
}
